test/sql/test_metadata.py

```diff
@@ -4780,11 +4853,190 @@ def test_schematype_ck_name_enum(self):
             ")",
         )
 
+    def test_schematype_ck_name_enum__no_ck(self):
+        """If a CheckConstraint is EXPLICITLY declared for a Enum column, it
+        should use the `ck_` prefix and not the `type_ck`.
+
+        If the `ck_` is missing in the naming_convention, then only the raw
+        constraint name is expected to be used."""
+        m1 = MetaData(
+            naming_convention={"type_ck": "type_ck_%(table_name)s_%(constraint_name)s",
+                               }
+        )
+        u1 = Table("user", m1, Column("x", Enum("a", "b", name="foo")))
+        # constraint is not hit
+        eq_(
+            [c for c in u1.constraints if isinstance(c, CheckConstraint)][
+                0
+            ].name,
+            "foo",
+        )
+        # but is hit at compile time
+        self.assert_compile(
+            schema.CreateTable(u1),
+            'CREATE TABLE "user" ('
+            "x VARCHAR(1), "
+            "CONSTRAINT foo CHECK (x IN ('a', 'b'))"
+            ")",
+        )
+
+    def test_schematype_ck_name_enum_not_on_name(self):
+        """If a CheckConstraint is IMPLICITLY created for a Enum column, it
+        should use the `type_ck_` prefix.
+
+        Ensure the correct naming_convention is chosen when `ck` and `type_ck`
+        are both present.
+        """
+        m1 = MetaData(
+            naming_convention={"ck": "ck_%(table_name)s_%(constraint_name)s",
+                               "type_ck": "type_ck_%(table_name)s_%(column_0_name)s",
+                               }
+        )
+        u1 = Table("user", m1, Column("x", Enum("a", "b")))
+        # constraint is not hit
+        eq_(
+            [c for c in u1.constraints if isinstance(c, CheckConstraint)][
+                0
+            ].name,
+            "_unnamed_",
```

Comment:
I hate the "_unnamed_" thing, can we maybe just do an all 1.4 commit here that whacks that whole "_defer_none_name" thing (assuming this is possible, I want to have a default "type_ck" convention to help with this) ?   I can work on that since it's my mess to clean up.

Reply:
1. I dislike the unnamed thing too.
2. I'm not sure `_defer_none_name` can be dropped.  if it can, I think it will be a lot of work and break many test cases.  a relevant test-case is the new one I added for your edge-case of columnless constraints -- `[test_schematype_ck_no_columns_on_constraint](https://github.com/sqlalchemy/sqlalchemy/pull/4792/files/6182721f8556a08485f1b089f24cf30f4705d1ed#diff-85c40f646c916a1455f90ca632061d28R5006-R5035)` (you may need to scroll all the way to the bottom; github is highlighting the code but not autoscrolling to the section)

recreated here...

            u1 = Table(
                "user", m1,
                Column("x", Integer),
                CheckConstraint('z > 10')
            )

This will automatically raise the error on `Table` - instead of `.compile` like the other cases - because the constraint name doesn't exist but it also isn't deferred.